### ISSUE TITLE
Feature Request: dns-01 should account for wildcard domains

### ISSUE THREAD

**Issue by willseward** [xenolf/lego#151]
Assuming someone had the domain `example.com` and `server1.example.com` were a CNAME for `example.com`, the dns-01 certificate request would fail. This is the case for topologies using L7 load balancing. 

After skimming the ACME protocol, I don't see any prohibition for looking for `server1.example.com` zone first, then after failing to find it, looking for  `example.com`, and so on... Correct me if I'm wrong.

**Comment by willseward** [comment]
Something like this: 57850777a37e200910b8e26c75fa3b1271742a53

I'm not a go programmer, so I'm hesitant to put in a PR

**Comment by xenolf** [comment]
Hey @willseward! I'm pretty sure we support CNAMES when checking for a live record. The support was added in #116. We are checking if the DNS record we request a cert for is in fact a CNAME and if it is, we use the delegated record for our logic. (see [here](https://github.com/xenolf/lego/blob/master/acme/dns_challenge.go#L91)) 

If I misunderstood the problem you are describing, please correct me :worried:

**Comment by willseward** [comment]
That commit may solve the problem on other providers, but the DO provider is still broken with respect to this issue. If you don't have the dns setup as a zone in DO, it will return a 404 and fail. It would be an easy fix to expand the Present method to this `Present(domain, authZone, token, keyAuth string)`, so that DO would know where to place the TXT record.

**Comment by xenolf** [comment]
Oh I'm sorry - I didn't understand this as a DO problem at all as you weren't mentioning it. So if I'm getting it straight now, the issue is that the DO API returns a 404 if the validation domain is not a zone but merely belongs to a wildcard CNAME?

Changing `Present` would not help with this I'm afraid as the code I linked to actually runs after the provider added the TXT record to validate it. 

Are setups like *.example.com CNAME to examplestatic.com also feasible? Because simply walking up the domain tree wouldn't catch this.

**Comment by willseward** [comment]
As long as example.com is the authoritative zone for *.example.com, placing a TXT record in example.com should validate successfully. If the CNAME for *.example.com points to examplestatic.com, I don't think that makes much of a difference.

I made some changes to the dns-01 test so that the provider will create a TXT record for the validation domain, but adds the record to the authoritative zone (hence the `authZone` in my previous comment) and everything seems to works fine.

Like this:

```go
zone, zerr := findZoneByFqdn(toFqdn(domain), recursiveNameserver)
if zerr != nil {
	log.Printf("Zone for '%s' could not be determined", domain)
	return zerr
}
log.Printf("Zone for domain is '%s'", zone)

dnsZone := unFqdn(zone)

// Generate the Key Authorization for the challenge
keyAuth, err := getKeyAuthorization(chlng.Token, s.jws.privKey)
if err != nil {
	return err
}

err = s.provider.Present(dnsZone, chlng.Token, keyAuth)
if err != nil {
	return fmt.Errorf("Error presenting token %s", err)
}
defer func() {
	err := s.provider.CleanUp(dnsZone, chlng.Token, keyAuth)
	if err != nil {
		log.Printf("Error cleaning up %s %v ", domain, err)
	}
}()
```

Again, excuse my go...

**Comment by janeczku** [comment]
@xenolf It's a bug in the DO provider. Any certificate request for a subdomain will fail because the `Provide()` method tries to create the validation record in the non-existent zone `subdomain.example.com` instead of in `example.com`. 
See https://github.com/xenolf/lego/blob/master/providers/dns/digitalocean/digitalocean.go#L53
/cc @mholt

**Comment by mholt** [comment]
Oops. I won't have time to fix that in the near future. Looks like @willseward probably has a good start on it.

**Comment by willseward** [comment]
I'd be more than happy to fix it

--
Wills Ward

**Comment by willseward** [comment]
https://github.com/xenolf/lego/pull/159

I might have refactored more than I should have. Let me know what you think.

**Comment by xenolf** [comment]
Alright :)

As the scope of this problem is only the DO provider, we should fix it in there.
@willseward If you want to tackle this then please open a PR with your changes and we go from there.

**Comment by xenolf** [comment]
Oh wow, what timing :)

**Comment by willseward** [comment]
Well, I can submit a PR with just DO changes if necessary.

**Comment by xenolf** [comment]
This was fixed by #176
